docs/presentations/linkedin-meetup-2024-04-24.qmd

```diff
@@ -0,0 +1,725 @@
+graph LR
+  A[data] --> B[batch processing] & C[stream processing] --> D[downstream]
+```
+
+## In the machine learning world...
+
+```{mermaid}
+graph TB
+  proddata --> sampled
+  model --> prodpipeline
+  subgraph "local env"
+  sampled[sampled data] --> local[local experimentation]
+  local <--> iterate
+  local --> model[finally, we have a production-ready model!]
+  end
+  subgraph "prod env"
+  proddata[production data] --> prodpipeline[production pipelines]
+  end
+```
+
+## {#local-experimentation-to-production-image}
+
+[Insert image here]
+
+## In the machine learning world...
+
+```{mermaid}
+graph TB
+  proddata --> sampled
+  model -- "code rewrite" --> prodpipeline
+  linkStyle 1 color:white;
+  subgraph "local env"
+  sampled[sampled data] --> local[local experimentation]
+  local <--> iterate
+  local --> model[finally, we have a production-ready model!]
+  end
+  subgraph "prod env"
+  proddata[production data] --> prodpipeline[production pipelines]
+  end
+```
+
+## A real-world example
+
+:::: {.columns}
+
+::: {.column width="50%"}
+pandas
+```python
+return (
+    clicks_df
+    .groupby(["user"])
+    .rolling("1h")
+    .size()
+)
+```
+:::
+
+::: {.column width="50%"}
+Flink SQL
+
+```sql
+SELECT
+  user,
+  COUNT(url) OVER (
+    PARTITION BY user
+    ORDER BY click_time
+    RANGE BETWEEN INTERVAL '1' HOUR PRECEDING AND CURRENT ROW
+  ) AS one_hour_user_click_cnt
+FROM clicks
+```
+:::
+
+::::
+
+## Code rewrites
+
+- From batch to streaming
+- From local experimentation to production
+- Backfilling a streaming feature on a batch backing table
+- …
+
+## The solution...
+
+Stream-batch unified API
+
+- Flink SQL
+- Spark DataFrame API
+- …
+
+## Stream-batch unification {.nostretch}
+
+:::: {.columns}
+
+::: {.column width="50%"}
+pandas
+```python
+return (
+    clicks_df
+    .groupby(["user"])
+    .rolling("1h")
+    .size()
+)
+```
+:::
+
+::: {.column width="50%"}
+Flink SQL
+```sql
+SELECT
+  user,
+  COUNT(url) OVER (
+    PARTITION BY user
+    ORDER BY click_time
+    RANGE BETWEEN INTERVAL '1' HOUR PRECEDING AND CURRENT ROW
+  ) AS one_hour_user_click_cnt
+FROM clicks
+```
+:::
+::::
+
+![](./arrow.png){fig-align="center" width=5%}
+
+Ibis
+```python
+agged = clicks.select(
+    _.user,
+    one_hour_user_click_cnt=_.url.count().over(
+        range=(-ibis.interval(hour=1), 0),
+        group_by=_.user,
+        order_by=_.click_time,
+    ),
+)
+```
+
+## Challenges of creating a unified API
+
+- Streaming is different
+  - Time semantics
+  - Long-running queries
+  - Sources and sinks
+  - …
+- Less established standards in streaming syntax
+
+## Ibis streaming today
+
+- Flink backend and RisingWave backend launched in Ibis 8.0
+- Introduction of watermarks, windowed aggregations, etc in Ibis
+
+## What's next?
+
+- Expand support of streaming operations and syntax
+- Continuously iterate on a stream-batch unified API
+- More streaming backends (Spark Structured Streaming)
+
+## Try it out now!
```

Comment:
this is cutoff at the bottom -- perhaps separate this into 2 slides, one on the browser REPL w/ a screenshot + one w/ installation instructions + the initial example?

<img width="1497" alt="image" src="https://github.com/ibis-project/ibis/assets/54814569/9e8bb901-a383-47fa-ab75-baaf69f28b03">

Reply:
this is fixed